### PR TITLE
Add updated cbf

### PR DESCRIPTION
**Pull request by Federico-PizarroBejarano**
Adding Lukas's CBF in the current safety filter format. 

I split the base CBF from the neural network CBF, and Allen's code can also extend the base CBF.

### CODE REVIEW

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+            'current_state_var': current_state_var,
+            'uncertified_action_var': uncertified_action_var,
+            'cost': cost
+        }
+
+    def solve_optimization(self,
+                           current_state,
+                           uncertified_action,
+                           ):
+        '''Solve the MPC optimization problem for a given observation and uncertified input.
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:166]: This is not MPC

File: experiments/cbf/config_overrides/cartpole_config.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+task_config:
+  seed: 42
+  ctrl_freq: 15
```
Comment (lukasbrunke) [inline experiments/cbf/config_overrides/cartpole_config.yaml:3]: This seems low? Do you see any effects from discretization?

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+        Args:
+            env_func (partial BenchmarkEnv): Functionalized initialization of the environment.
+            slope (float): The slope of the linear function in the CBF.
+            soft_constrainted (bool): Whether to use soft or hard constraints.
+            slack_weight (float): The weight of the slack in the optimization.
+            slack_tolerance (float): How high the slack can be in the optimization.
+        '''
+
+        # TODO: Combine with CLF approach for stabilization
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:37]: Should we remove these todos? And open new issues for each of these to track?

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        self.X = self.model.x_sym
+        self.u = self.model.u_sym
+
+        if self.env.NAME == 'cartpole':
+            self.state_limits = [min(abs(self.state_constraint.upper_bounds[i]), abs(self.state_constraint.lower_bounds[i])) for i in range(self.model.nx)]
+            self.cbf = cbf_cartpole(self.X, self.state_limits)
+        else:
+            raise NotImplementedError('[Error] Currently CBF is only implemented for the cartpole system.')
+
+        # TODO: Extend this to systems that are not control affine. Then we would need to move away from a QP solver
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:76]: See comment above about todos

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+            _, success = self.certify_action(state, control_input)
+
+            if not success:
+                infeasible_states.append(state)
+                num_infeasible += 1
+                barrier_at_x = self.cbf(X=state)['cbf']
+
+                # Check if the infeasible point is inside or outside the superlevel set. Note that the sampled region makes up a
+                # box, but the superlevel set is not. The superlevel set only needs to be contained inside the box.
+                if barrier_at_x > 0:
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:290]: Maybe something like `barrier_at_x >= 0.0 + eps` where eps is some positive number is more explicit?

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+            'opti': opti,
+            'u_var': u_var,
+            'slack_var': slack_var,
+            'current_state_var': current_state_var,
+            'uncertified_action_var': uncertified_action_var,
+            'cost': cost
+        }
+
+    def solve_optimization(self,
+                           current_state,
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:163]: Add types? `current_state: np.ndarray` etc.

File: safe_control_gym/safety_filters/cbf/cbf.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                           uncertified_action,
+                           ):
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf.py:165]: Same for the output: `-> Tuple(np.ndarray, bool)`

File: safe_control_gym/safety_filters/cbf/cbf_nn.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+        opti.set_value(current_state_var, current_state)
+        opti.set_value(uncertified_action_var, uncertified_action)
+
+        torch_state = torch.from_numpy(current_state)
+        torch_state = torch.unsqueeze(torch_state, 0)
+        torch_state = torch_state.to(torch.float32)
+        a_b = self.mlp(torch_state)
+        a_b = a_b.detach().numpy()
+        a = a_b[0, :self.model.nu]
+        b = a_b[0, -1]
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf_nn.py:183]: We could turn this into another method. We reuse this later as well

File: safe_control_gym/safety_filters/cbf/cbf_nn.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+                lie_derivative_values[counter, :] = self.lie_derivative(X=obs, u=blended_input)['LfV']
+
+                # Determine the estimated Lie derivative
+                torch_state = torch.from_numpy(obs)
+                torch_state = torch.unsqueeze(torch_state, 0)
+                torch_state = torch_state.to(torch.float32)
+                a_b = self.mlp(torch_state)
+                a_b = a_b.detach().numpy()
+                a = a_b[0, :self.model.nu]
+                b = a_b[0, -1]
```
Comment (lukasbrunke) [inline safe_control_gym/safety_filters/cbf/cbf_nn.py:337]: These lines are the same as the above

**Review by lukasbrunke — CHANGES_REQUESTED** [review]
Made some minor suggestions.

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 1 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf.py]: Fixed

File: safe_control_gym/safety_filters/cbf/cbf_nn.py (hunk 8 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf_nn.py]: Done.

File: safe_control_gym/safety_filters/cbf/cbf_nn.py (hunk 9 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf_nn.py]: Done.

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 6 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf.py:163]: We haven't been making types explicit throughout the gym except for in the __init__ function really. I think this is a more general style discussion.

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 3 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf.py:37]: @JacopoPan general question on structure: would you rather leave the TODOs in (making it look somewhat unfinished), open an issue, or some other way of tracking these issues?

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 6 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/safety_filters/cbf/cbf.py:163]: @JacopoPan @Justin-Yuan idk if you want to move towards being more explicit on typing

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 3 above)
Comment (JacopoPan) [reply safe_control_gym/safety_filters/cbf/cbf.py:37]: Both TODOs and issues are good.

File: safe_control_gym/safety_filters/cbf/cbf.py (hunk 6 above)
Comment (JacopoPan) [reply safe_control_gym/safety_filters/cbf/cbf.py:163]: Type (and add defaults) wherever possible but I would NOT to make mandatory repo-wide

**Review by lukasbrunke — APPROVED** [review]
Looks great!

**Pull request merged**